### PR TITLE
Implement Extended Currency DataModel

### PR DESCRIPTION
**Pull request by younies**

### CODE REVIEW

File: components/experimental/src/dimension/provider/extended_currency.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_experimental::dimension::provider::extended_currency)
+)]
+#[yoke(prove_covariance_manually)]
+pub struct CurrencyExtendedDataV1<'data> {
+    /// A mapping from each currency's ISO code to its associated formatting patterns.
+    // Using CurrencyPatternConfig until implementing AsULE for ExtendedCurrencyPatternConfig.
+    // use short for long right now.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub patterns_config: ZeroMap<'data, Count, CurrencyPatternConfig>,
```
Comment (sffc) [inline components/experimental/src/dimension/provider/extended_currency.rs:45]: Suggestion: I think in general the best way to store things that vary by plural form is to pull out the `other` variant and store all the others in the map. Something more like

```rust
/// For plural form 'other'
pub pattern_config_other: CurrencyPatternConfig,
/// For all other plural forms and explicit forms
pub pattern_config_plurals: ZeroMap<'data, Count, CurrencyPatternConfig>,
```

Comment (younies) [reply components/experimental/src/dimension/provider/extended_currency.rs:45]: great idea, we treated it as default pattern.

File: provider/datagen/src/transform/cldr/currency/extended.rs [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+//             locale.set_aux();
+//             locale
+//         }));
+
+//         Ok(r)
+//     }
+// }
+
+impl IterableDataProviderInternal<CurrencyExtendedDataV1Marker> for crate::DatagenProvider {
```
[the remark this answers is not in the document]
Comment (younies) [reply provider/datagen/src/transform/cldr/currency/extended.rs:97]: https://github.com/unicode-org/icu4x/blob/daa13bb3583713b845d1a949aa8fb54d2f07b59f/provider/datagen/src/transform/cldr/datetime/neo_skeleton.rs#L168

File: provider/datagen/src/transform/cldr/currency/extended.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+        Ok(DataResponse {
+            metadata: Default::default(),
+            payload: Some(DataPayload::from_owned(data)),
+        })
+    }
+}
+
+impl IterableDataProvider<CurrencyExtendedDataV1Marker> for DatagenProvider {
+    fn supported_requests(&self) -> Result<HashSet<(DataLocale, DataKeyAttributes)>, DataError> {
```
Comment (younies) [inline provider/datagen/src/transform/cldr/currency/extended.rs:82]: I have two questions:

1. Shall we return the `HashSet` with the `DataLocale` and `DataKeyAttributes` without adding the auxiliary key to the `DataLocale` ?
2. How to create a `DataKeyAttributes` from a `TinyAsciiStr<3>`  which is the currency code ?

Comment (sffc) [reply provider/datagen/src/transform/cldr/currency/extended.rs:82]: @robertbastian

Comment (robertbastian) [reply provider/datagen/src/transform/cldr/currency/extended.rs:82]: 1. `DataKeyAttributes` are not in the `DataLocale` anymore. If you previously used `set_aux`, now just return the `DataKeyAttributes` separately
2. `DataKeyAttributes::from_tinystr(x.resize())`

**Comment by younies** [comment]
@sffc and @robertbastian 

when I run `cargo make testdata`

I received the following error message
```
thread 'provider::tests::make_testdata::make_testdata' panicked at provider/datagen/src/provider/tests/make_testdata.rs:302:9:
Expected violations list does not match found violations!
If the new list is smaller, please update EXPECTED_VIOLATIONS in make-testdata.rs
If it is bigger and that was unexpected, please make sure the key remains zero-copy, or ask ICU4X team members if it is okayto temporarily allow for this key to be allowlisted.
Expected:
[DataKey{datetime/skeletons@1}]
Found:
[DataKey{currency/extended@1}, DataKey{datetime/skeletons@1}]
Expected (transient):
[DataKey{list/and@1}, DataKey{list/or@1}, DataKey{list/unit@1}]
Found (transient):
[DataKey{list/and@1}, DataKey{list/or@1}, DataKey{list/unit@1}]
```

**Comment by sffc** [comment]
It means that `DataKey{currency/extended@1}` contains allocated data and you need to make it zero-copy.

**Comment by sffc** [comment]
It looks like you forgot to add `#[cfg_attr(feature = "serde", serde(borrow))]` on all of the fields. If you don't have that annotation, they always get deserialized as fully owned (allocated).

**Comment by younies** [comment]
> It looks like you forgot to add `#[cfg_attr(feature = "serde", serde(borrow))]` on all of the fields. If you don't have that annotation, they always get deserialized as fully owned (allocated).

@sffc : thanks for the notice, but even I applied this, still I get the same message

```
Expected violations list does not match found violations!
If the new list is smaller, please update EXPECTED_VIOLATIONS in make-testdata.rs
If it is bigger and that was unexpected, please make sure the key remains zero-copy, or ask ICU4X team members if it is okayto temporarily allow for this key to be allowlisted.
Expected:
[DataKey{datetime/skeletons@1}]
Found:
[DataKey{currency/extended@1}, DataKey{datetime/skeletons@1}]
Expected (transient):
[DataKey{list/and@1}, DataKey{list/or@1}, DataKey{list/unit@1}]
Found (transient):
[DataKey{list/and@1}, DataKey{list/or@1}, DataKey{list/unit@1}]
```

**Comment by robertbastian** [comment]
Serde is too limited to borrow through an `Option`, you need to use one of the methods in `provider/core/src/serde/borrow_de_utils.rs`

**Comment by sffc** [comment]
https://github.com/serde-rs/serde/issues/2016 again

**Comment by robertbastian** [comment]
Wow they locked that thread.

**Comment by sffc** [comment]
Since the Serde issue doesn't keep track of linked issues any more, let's start using https://github.com/unicode-org/icu4x/pull/1556 to keep track of all the cases where we've wasted time re-discovering and re-working-around this issue (which I still hold to be fully in scope for Serde core)

**Comment by younies** [comment]
right now, I added those fields to the display_names map in order to avoid this :)

**Comment by sffc** [comment]
This looks good, but I would like to see a way to reduce the number of currency JSON files generated in testdata if possible? Thoughts @robertbastian?

**Comment by younies** [comment]
@sffc, I have reduced the currencies files.

File: components/experimental/src/dimension/provider/extended_currency.rs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+#[yoke(prove_covariance_manually)]
+pub struct CurrencyExtendedDataV1<'data> {
+    // TODO: Implement currency pattern selection logic to choose between standard or standard next to text pattern.
+    /// Contains the localized display names for a currency based on plural rules.
+    /// For instance, in the "en" locale for the "USD" currency:
+    ///     - "US Dollars" when count is `zero`,
+    ///     - "US Dollar" when count is `one`,
+    ///     ... etc.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub display_names: ZeroMap<'data, Count, str>,
```
Comment (sffc) [inline components/experimental/src/dimension/provider/extended_currency.rs:43]: Issue: As described in [the spec](https://unicode.org/reports/tr35/tr35-numbers.html#Currencies), you also need to include `unitPattern` along with the display names. Consider adding it to currency essentials.

Comment (younies) [reply components/experimental/src/dimension/provider/extended_currency.rs:43]: The patterns are already in currency essentials. I just need to add the selector in the next PR to choose the standard or standard-alphaNextToNumber

Comment (sffc) [reply components/experimental/src/dimension/provider/extended_currency.rs:43]: I think unitPattern is different than standard and standard-alphaNextToNumber. It is a third pattern. 

But you can add it in the next PR. This one is fine.

**Review by sffc — APPROVED** [review]

**Pull request merged**